libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,316 @@
+// cache.
+func newDiskBlockCacheStandardFromStorage(config diskBlockCacheConfig,
+	blockStorage, lruStorage storage.Storage, maxBytes uint64) (
+	*DiskBlockCacheStandard, error) {
+	log := config.MakeLogger("KBC")
+	blockDb, err := openLevelDB(blockStorage)
+	if err != nil {
+		return nil, err
+	}
+
+	lruDb, err := openLevelDB(lruStorage)
+	if err != nil {
+		blockDb.Close()
+		return nil, err
+	}
+	return &DiskBlockCacheStandard{
+		config:   config,
+		maxBytes: maxBytes,
+		log:      log,
+		blockDb:  blockDb,
+		lruDb:    lruDb,
+	}, nil
+}
+
+// newDiskBlockCacheStandard creates a new *DiskBlockCacheStandard with a
+// specified directory on the filesystem as storage.
+func newDiskBlockCacheStandard(config diskBlockCacheConfig, dirPath string,
+	maxBytes uint64) (*DiskBlockCacheStandard, error) {
+	blockDbPath := filepath.Join(dirPath, blockDbFilename)
+	blockStorage, err := storage.OpenFile(blockDbPath, false)
+	if err != nil {
+		return nil, err
+	}
+	lruDbPath := filepath.Join(dirPath, lruDbFilename)
+	lruStorage, err := storage.OpenFile(lruDbPath, false)
+	if err != nil {
+		blockStorage.Close()
+		return nil, err
+	}
+	return newDiskBlockCacheStandardFromStorage(config, blockStorage,
+		lruStorage, maxBytes)
+}
+
+// TODO: Fix getSizesLocked(), where leveldb.DB.SizeOf() currently doesn't work
+// for the full range. As it is right now, this is a very cheap operation,
+// since it just takes the beginning and end offsets from levelDB and returns
+// the difference.
+func (cache *DiskBlockCacheStandard) getSizesLocked() (blockSize int64,
+	lruSize int64, err error) {
+	// a util.Range{nil, nil} is supposed to work for the full range based on
+	// other methods, but it appears not to work for SizeOf. Need to implement
+	// by querying the files in the LevelDB directory instead. Ugh.
+	blockSizes, err := cache.blockDb.SizeOf([]util.Range{{}})
+	if err != nil {
+		return -1, -1, err
+	}
+	lruSizes, err := cache.lruDb.SizeOf([]util.Range{{}})
+	if err != nil {
+		return -1, -1, err
+	}
+	return blockSizes[0], lruSizes[0], nil
+}
+
+// compactCachesLocked manually forces both the block cache and LRU cache to
+// compact their underlying data.
+func (cache *DiskBlockCacheStandard) compactCachesLocked(ctx context.Context) {
+	cache.blockDb.CompactRange(util.Range{})
+	cache.lruDb.CompactRange(util.Range{})
+	bSize, lruSize, err := cache.getSizesLocked()
+	cache.log.CDebugf(ctx, "Disk Cache bSize=%d lruSize=%d err=%+v", bSize, lruSize, err)
+}
+
+// lruKey generates an LRU cache key from a tlf.ID and a binary-encoded block
+// ID.
+func (*DiskBlockCacheStandard) lruKey(tlfID tlf.ID, blockKey []byte) []byte {
+	return append(tlfID.Bytes(), blockKey...)
+}
+
+// updateLruLocked updates the LRU time of a block in the LRU cache to
+// time.Now().
+func (cache *DiskBlockCacheStandard) updateLruLocked(tlfID tlf.ID,
+	blockKey []byte) error {
+	key := cache.lruKey(tlfID, blockKey)
+	val, err := time.Now().MarshalBinary()
```

Comment:
This makes it hard to test reliably -- any chance you'd consider using config.Clock()?  The downside being it's one more thing to add to the config.